### PR TITLE
moves coauthored notes to user model (resolves #1793)

### LINKED ISSUE #1793: Refactoring coauthored notes query into a method in User.rb
Hi, this is a [first-timers-only issue](https://publiclab.github.io/community-toolbox/). This means we've worked to make it more legible to folks who either haven't contributed to our codebase before, or even folks who haven't contributed to open source before.

If that's you, we're interested in helping you take the first step and can answer questions and help you out as you do. Note that we're especially interested in contributions from people from groups underrepresented in free and open source software!

We know that the process of creating a pull request is the biggest barrier for new contributors. This issue is for you 💝

If you have contributed before, consider leaving this one for someone new, and looking through our general help wanted issues. Thanks!

🤔 What you will need to know.

Nothing. This issue is meant to welcome you to Open Source :) We are happy to walk you through the process.

📋 Step by Step

- [ ]  🙋 Claim this issue: Comment below. If someone else has claimed it, ask if they've opened a pull request already and if they're stuck -- maybe you can help them solve a problem or move it along!
- [ ]  📝 Update the file [https://github.com/publiclab/plots2/blob/master/app/views/dashboard/_node_meta.html.erb](https://github.com/publiclab/plots2/blob/master/app/views/dashboard/_node_meta.html.erb) in the plots2 repository (press the little pen Icon) and edit the line as shown below. [See this page](https://github.com/publiclab/plots2/projects/2) for some help in taking your first steps!
- [ ]   💾 Commit your changes
- [ ]  🔀 Start a Pull Request. There are two ways how you can start a pull request:

### PROBLEM :

We need to Refactor the Coauthored notes query (soon to be used on profile pages [like this](https://publiclab.org/profile/warren) to display posts an author has assisted with) into a method in the `User.rb` model (`/app/models/user.rb`) and to write a Unit test for it (in `/test/unit/user_test.rb`).

### SOLUTION :

The query is given here :
https://github.com/publiclab/plots2/blob/de29744270e6ea518e6c69bb47827b51125e6fc7/app/controllers/users_controller.rb#L111

The above query should be moved into the `user.rb` model as :  

```ruby
def coauthored_notes
    # the query goes here 
end
```

And the test for the above query goes here -- you can see other examples of tests in this file too: 
[test/unit/user_test.rb](https://github.com/publiclab/plots2/blob/master/test/unit/user_test.rb)

## Thank you!

Your help makes Public Lab better! We *deeply* appreciate your helping refine and improve this site. 

To learn how to write really great issues, which increases the chances they'll be resolved, see:

https://publiclab.org/wiki/developers#Contributing+for+non-coders

### PR DESCRIPTION
**Pull request by zhirschtritt**
This pull is in progress

- [x] Create `coauthored_notes` method in User.rb and use in user_controller
- [x] Write unit tests for method

In reference to issue #1793

### CODE REVIEW

**Comment by jywarren** [comment]
Awesome, this looks good. Are you familiar with rails testing? It would be great to add a unit test in /test/unit/user_test.rb which adds a co-authored tag like "with:bob" to an existing note (not originally authored by bob) and then `assert` that Bob has that note listed when we call ". coauthored_notes". Would u be up for trying that? I can help more with this too - and there are good examples of tests already in that file.

**Comment by zhirschtritt** [comment]
Hmm, might need a hand with this. Here's what I'm thinking: 

```
    jeff = rusers(:jeff)
    bob = rusers(:bob)
    assert bob.coauthored_notes.empty?

    jeffs_note = node(:one)
    jeffs_note.tag.create(name: "with:bob")

    coauthored_note = bob.coauthored_notes.first

    assert_not_nil coauthored_note
    assert_equal jeffs_note, coauthored_note
```
This does not work, but I feel like I'm not creating a

**Comment by jywarren** [comment]
If you push up your latest code we can take a look at exactly how/where it
fails. Looks like a great start! I think the `_note.tag.create()` section
may need work; there are examples of creating tags elsewhere in the tests
you could use as a starting point.

Cool!

File: test/unit/user_test.rb [hunk 1]
```diff
@@ -83,4 +83,18 @@ class UserTest < ActiveSupport::TestCase
...
      assert_equal [node2], lurker.content_followed_in_past_period(2.hours.ago)
   end
+
+  test 'returns nodes coauthored by user with coauthored_notes method' do
+    jeff = rusers(:jeff)
+    bob = rusers(:bob)
+    assert bob.coauthored_notes.empty?
+
+    jeffs_note = node(:one)
+    jeffs_note.tag.create(name: "with:bob")
```
Comment (jywarren) [inline test/unit/user_test.rb]: See this example:

https://github.com/publiclab/plots2/blob/a91d7faf2b74e928c1013e1bcd1f4816c5cbf9e3/test/integration/node_insert_extras_test.rb#L21

**Comment by jywarren** [comment]
Wow, this looks awesome! I'm working on an upgrade to Rails 4.1 right now, so I'm going to try to wrap that up ASAP before merging this in, to avoid conflicts. I'll help resolve any if they occur. 

Thanks a lot, @zhirschtritt !!! This is fantastic.

**Comment by jywarren** [comment]
OK, #1799 is finally merged in! I'll circle back to this to see if we can resolve the conflicts now. Sorry that took much longer than I thought!!!

**Comment by jywarren** [comment]
OK, i've done a rebase over the chaos I introduced in the Rails 4.1 upgrade, preserving your changes. I hope i got it right!

**Comment by jywarren** [comment]
Hooray! Done, and merging. Thanks for your patience, and for your help, Zach! Hope it went well for you and if you want, there are lots more issues to tackle!

https://publiclab.github.io/community-toolbox/

**Pull request merged**